File: database/src/main/kotlin/br/com/dillmann/nginxsidewheel/database/common/DatabaseStartup.kt
package br.com.dillmann.nginxsidewheel.database.common

import br.com.dillmann.nginxsidewheel.core.common.log.logger
import br.com.dillmann.nginxsidewheel.core.common.provider.ConfigurationProvider
import br.com.dillmann.nginxsidewheel.core.common.startup.StartupCommand
import org.jetbrains.exposed.sql.Database

internal class DatabaseStartup(configurationProvider: ConfigurationProvider): StartupCommand {
    private val configurationProvider = configurationProvider.withPrefix("nginx-sidewheel.database")
    override val priority = 0

    override suspend fun execute() {
        val url = configurationProvider.get("url")
        val username = configurationProvider.get("username")
        val password = configurationProvider.get("password")

        logger<DatabaseStartup>().info("Starting database connection to $url with username $username")

        Database.connect(
            url = url,
            user = username,
            password = password,
        )
    }
}

File: database/database/database_startup.go
package database

import (
	"dillmann.com.br/nginx-ignition/core/common/lifecycle"
)

type startup struct {
	database *Database
}

func Register(lifecycle *lifecycle.Lifecycle, database *Database) {
	command := &startup{database}
	lifecycle.RegisterStartup(command)
}

func (d startup) Priority() int {
	return 0
}

func (d startup) Async() bool {
	return false
}

func (d startup) Run() error {
	return d.database.init()
}
